Support CCS minimize round trips in async search

quux00 opened the pull request:
This commit makes the smallest set of changes to allow async-search based cross-cluster search
to work with the CCS minimize_round_trips feature without changing the internals/architecture of
the search action.

When ccsMinimizeRoundtrips is set to true on SubmitAsyncSearchRequest, the AsyncSearchTask on the
primary CCS coordinator sends a synchronous SearchRequest to all to clusters for a remote coordinator
to orchestrate and return the entire result set to the CCS coordinator as a single response.

This is the same functionality provided by synchronous CCS search using minimize_round_trips.
However, since this is an async search, it means that the async search coordinator has no visibility
into search progress on the remote clusters while they are running the search, thus losing one of
the key features of async search. However, this is a good first approach for improving overall search
latency for cross cluster searches that query a large number of shards on remote clusters, since
Kibana does not currently expose incremental progress of an async search to users.

quux00 committed:
Support CCS minimize round trips in async search

This commit makes the smallest set of changes to allow async-search based cross-cluster search
to work with the CCS minimize_round_trips feature without changing the internals/architecture of
the search action.

When ccsMinimizeRoundtrips is set to true on SubmitAsyncSearchRequest, the AsyncSearchTask on the
primary CCS coordinator sends a synchronous SearchRequest to all to clusters for a remote coordinator
to orchestrate and return the entire result set to the CCS coordinator as a single response.

This is the same functionality provided by synchronous CCS search using minimize_round_trips.
However, since this is an async search, it means that the async search coordinator has no visibility
into search progress on the remote clusters while they are running the search, thus losing one of
the key features of async search. However, this is a good first approach for improving overall search
latency for cross cluster searches that query a large number of shards on remote clusters, since
Kibana does not currently expose incremental progress of an async search to users.

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

javanna commented on qa/multi-cluster-search/src/test/java/org/elasticsearch/search/CCSDuelIT.java:
```diff
@@ -789,20 +832,144 @@ private static void duelSearch(SearchRequest searchRequest, Consumer<SearchRespo
+        try {
+            final String responseId = fanOutResponse.getId();
+            assertBusy(() -> {
+                AsyncSearchResponse resp = getAsyncSearch(responseId, restHighLevelClient.getParserConfig());
```
out of curiosity: you could have cut corners here, and set a high enough timeout for submitAsyncSearch so that you never have to call get and delete. What made you go this route instead?

quux00 replied:
Since users doing CCS will frequently have queries that run longer than 1s, I thought it best to include retrieving the results from the GET endpoint rather than a synchronous return of the POST call.

Ideally I'd like the test to sometimes return synchronously (within the timeout) and sometimes run longer than the timeout, but that would add even more complexity to the test, so I didn't varying the time.

javanna commented on server/src/main/java/org/elasticsearch/action/search/SearchProgressListener.java:
```diff
@@ -32,6 +32,15 @@ public abstract class SearchProgressListener {
     private List<SearchShard> shards;
 
+    /**
+     * Executed one when minimize round trips is enabled to indicate the number of remote clusters
```
The beginning of the sentence has something off :)

quux00 replied:
Fixed.

javanna commented on server/src/main/java/org/elasticsearch/action/search/SearchProgressListener.java:
```diff
@@ -32,6 +32,15 @@ public abstract class SearchProgressListener {
+     * @param hasLocalShards whether the search will include searching local shards
+     * @param numRemoteClusters how many remote clusters are part of the CCS search
+     */
+    protected void onMinimizeRoundtrips(boolean hasLocalShards, int numRemoteClusters) {}
```
one thing I was wondering is if we need an additional callback, or we could rely on the existing onListShards. One reason why we may need it is that there is a case where you only search to a single remote cluster and no local indices, in which case I am not even sure onListShards is called. Maybe you also found other reasons.

quux00 replied:
Yeah, that's why I marked it as WIP. I wanted to do a PR now to see if it passes the build (I forgot that I could do a draft PR until after I created it). I still have more research, testing and experimentation to do before going for a full PR review. Thanks for the early feedback!

javanna replied:
sounds good, this looks already pretty close, another aspect I would look into is whether we need additional testing or the duel test is enough.

quux00 replied:
> One reason why we may need it is that there is a case where you only search to a single remote cluster and no local indices
I found a bug in this code for that exact case - where only querying a remote cluster. In that case, a call to notifyListShards was still necessary to initialize the searchResponse (MutableSearchResponse) in [AsyncSearchTask](https://github.com/elastic/elasticsearch/pull/96012/files#diff-24921eca27bf57c7b3d8d3ba7f02acbf6449d91e43e49bbd9df6377843adb19aR461).

So I've fixed that and added tests to CCSDuelIT that also tests a remote-cluster only (no local indices), which uncovers this bug and it now passes.

javanna reviewed (commented):
looks good to me so far, I am curious on why it's marked WIP ;) maybe we should add some indication in the get async search API response that some remote clusters are minimizing roundtrips?

quux00 committed:
Added logic into SearchResponse constructor to increment the successful count in Clusters - for async minimizeRoundtrips usage

quux00 committed:
Remove debugging comments and logging

quux00 committed:
Created hasLocalIndicesToSearch in TransportSearchAction to handle TODO action I left

quux00 committed:
Logging added to trace when/where MutableSearchResponse creates a new SearchResponse

quux00 committed:
Added SearchClustersInfo wrapper to Clusters in MutableSearchResponse

MutableSearchResponse now has logic in buildResponse about whether to
update the Clusters object to indicate that the local cluster has finished.

quux00 committed:
code cleanup after completing work to move Cluster update logic to MutableSearchResponse

quux00 committed:
Removed onMinimizeRoundtrips and notifyMinimizeRoundtrips from SearchProgressListener.

Modified SearchResponse.Clusters to optionally track remoteClusters and ccsMinimizeRoundtrips.

As the original Clusters object was intended to be immutable, I've added a "frozen" flag
that is set when the original constructor is called.

CCS AsyncSearch minimizeRoundTrips uses this new Clusters variant.

onListShards, with the revised impl and usage of SearchResponse.Clusters, now handles
notifying the SearchProgressListener in the AsyncSearchTask about the number of remote
clusters, whether local shards are being searched and ccsMinimizeRoundtrips status.

quux00 committed:
Changed SearchResponse.Clusters back to an immutable object.

Made the Clusters and localClusterStatusUpdated fields to be updated in a synchronized block.
in MutableSearchResponse.SearchClustersInfo to make them safe for multi threaded access.

quux00 commented:
With these changes, we have removed the notifyMinimizeRoundtrips/onMinimizeRoundtrips call to the SearchProgressListener in favor of modifying the `SearchResponse.Cluster` object to hold the information needed.

The `_clusters` object in the REST response for async_search with minimizeRoundTrips starts out at "successful: 0" and can move to "successful: 1" if there is a local cluster search that finishes while one or more of the remote clusters are still doing the search. 

Here's an example of the progression in the `_clusters` output for a CCS search using minimize-roundtrips against three 3-node clusters:

```javascript
POST blogs,remote1:*,remote2:*/_async_search?ccs_minimize_roundtrips=true
{
   // ... search body for a slow search not shown
}

// after _async_seach submit
{
  "id": "FkF3RE1KZnFSU3M2VUdsWEZjVnlmZFEaQVRnZDdaNjBUZ1N0RU1sUjl0QmphUTo2OTI=",
  "is_partial": true,
  "is_running": true,
  "start_time_in_millis": 1684516160099,
  "expiration_time_in_millis": 1684948160099,
  "response": {
    "took": 806,
    "timed_out": false,
    "terminated_early": false,
    "num_reduce_phases": 0,
    "_shards": {
      "total": 3,
      "successful": 0,  // no shards completed
      "skipped": 0,
      "failed": 0
    },
    "_clusters": {
      "total": 3,
      "successful": 0,  // no clusters completed
      "skipped": 0
    },
    "hits": {
      "total": {
        "value": 0,
        "relation": "gte"
      },
      "max_score": null,
      "hits": []
    }
  }
}

// while running (local cluster has finished; two remote cluster searches still running
GET /_async_search/FkF3RE1KZnFSU3M2VUdsWEZjVnlmZFEaQVRnZDdaNjBUZ1N0RU1sUjl0QmphUTo2OTI

{
  "id": "FkF3RE1KZnFSU3M2VUdsWEZjVnlmZFEaQVRnZDdaNjBUZ1N0RU1sUjl0QmphUTo2OTI=",
  "is_partial": true,
  "is_running": true,
  "start_time_in_millis": 1684516160099,
  "expiration_time_in_millis": 1684948160099,
  "response": {
    "took": 5973,
    "timed_out": false,
    "terminated_early": false,
    "_shards": {
      "total": 3,
      "successful": 3,  // local shards completed
      "skipped": 0,
      "failed": 0
    },
    "_clusters": {
      "total": 3,
      "successful": 1, // local cluster has completed
      "skipped": 0
    },
    "hits": {
      "total": {
        "value": 142,
        "relation": "eq"
      },
      "max_score": null,
      "hits": []
    }
  }
}


// when all clusters completed
GET /_async_search/FkF3RE1KZnFSU3M2VUdsWEZjVnlmZFEaQVRnZDdaNjBUZ1N0RU1sUjl0QmphUTo2OTI

{
  "id": "FkF3RE1KZnFSU3M2VUdsWEZjVnlmZFEaQVRnZDdaNjBUZ1N0RU1sUjl0QmphUTo2OTI=",
  "is_partial": false,
  "is_running": false,
  "start_time_in_millis": 1684516160099,
  "expiration_time_in_millis": 1684948160099,
  "response": {
    "took": 10350,
    "timed_out": false,
    "num_reduce_phases": 4,
    "_shards": {
      "total": 9,
      "successful": 9,
      "skipped": 0,
      "failed": 0
    },
    "_clusters": {
      "total": 3,
      "successful": 3,
      "skipped": 0
    },
    "hits": {
      "total": {
        "value": 432,
        "relation": "eq"
      },
      "max_score": 17.110811,
      "hits": [
        ...
```

Note that this is NOT the behavior with ccs_minimize_roundtrips=false. There there _clusters output is always the final output where total == successful:

```
    "_clusters": {
      "total": 3,
      "successful": 3,
      "skipped": 0
    },
```

quux00 commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java:
```diff
@@ -147,6 +177,18 @@ void addQueryFailure(int shardIndex, ShardSearchFailure shardSearchFailure) {
 
     private SearchResponse buildResponse(long taskStartTimeNanos, InternalAggregations reducedAggs) {
+        if (clustersInfo.getClusters() != null && clustersInfo.isLocalClusterStatusUpdated() == false) {
+            synchronized (this) {
```
Does this section need to be thread safe? Can `buildResponse` be called by two threads on the same `MutableSearchReponse` object? (Since other methods in this class are `synchronized`, I'm guessing the answer is yes, but want to double check with the reviewers.)

javanna replied:
yes, but buildResponse is only called from synchronized methods, so no additional synchronization is needed?

quux00 commented:
I tested a case where the remote clusters are down:

```javascript
// case where skip_unavailable: true is NOT set for a remote cluster and it is down:
POST blogs,remote1:*,remote2:*/_async_search?ccs_minimize_roundtrips=true
...

// response:
{
  "id": "FjJJZG9yWUhOU3VLaTJzNXZMalZaTkEaaVhraVNDT2VTenFoUDlod21hYmlKZzo5MTE=",
  "is_partial": true,
  "is_running": false,
  "start_time_in_millis": 1684767109136,
  "expiration_time_in_millis": 1685199109136,
  "response": {
    "took": 10927,
    "timed_out": false,
    "terminated_early": false,
    "_shards": {
      "total": 3,
      "successful": 3,
      "skipped": 0,
      "failed": 0
    },
    "_clusters": {
      "total": 3,
      "successful": 1,
      "skipped": 0
    },
    "hits": {
      "total": {
        "value": 167,
        "relation": "eq"
      },
      "max_score": null,
      "hits": []
    }
  },
  "error": {
    "type": "status_exception",
    "reason": "error while executing search",
    "caused_by": {
      "type": "connect_transport_exception",
      "reason": "[][127.0.0.1:9301] connect_exception",
      "caused_by": {
        "type": "uncategorized_execution_exception",
        "reason": "Failed execution",
        "caused_by": {
          "type": "execution_exception",
          "reason": "execution_exception: io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: /127.0.0.1:9301",
          "caused_by": {
            "type": "i_o_exception",
            "reason": "Connection refused: /127.0.0.1:9301",
            "caused_by": {
              "type": "i_o_exception",
              "reason": "Connection refused"
            }
          }
        }
      }
    }
  }
}

// where skip_unavailable:true
POST blogs,remote1:*,remote2:*/_async_search?ccs_minimize_roundtrips=true
...

// response:
{
  "id": "FnE0SjZrb2lOVFR5ejFCN3dtX0NHbVEbaVhraVNDT2VTenFoUDlod21hYmlKZzoxMDIz",
  "is_partial": false,
  "is_running": false,
  "start_time_in_millis": 1684767263845,
  "expiration_time_in_millis": 1685199263845,
  "response": {
    "took": 11943,
    "timed_out": false,
    "num_reduce_phases": 3,
    "_shards": {
      "total": 6,
      "successful": 6,
      "skipped": 0,
      "failed": 0
    },
    "_clusters": {
      "total": 3,
      "successful": 2,
      "skipped": 1   // was properly updated at the end (started off as 0)
    },
    "hits": {
      "total": {
        "value": 370,
        "relation": "eq"
      },
      "max_score": 16.595804,
      "hits": [
        {
...

javanna commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java:
```diff
@@ -109,12 +115,36 @@ synchronized void updatePartialResponse(
-        assert response.getSkippedShards() == skippedShards
-            : "received number of skipped shards differs from the one " + "notified through onListShards";
+
+        if (ccsMinimizeRoundtrips) {
```
I think you could make this a method that is called from an assert so you avoid doing the conditional when asserts are disabled.

javanna commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java:
```diff
@@ -165,10 +207,30 @@ private SearchResponse buildResponse(long taskStartTimeNanos, InternalAggregatio
+            || (clusters.getSuccessful() + clusters.getSkipped()) >= clusters.getTotal()) {
+            return false;
+        }
+        return totalShards == (successfulShards + skippedShards + failedShards);
```
I am afraid it is a bit delicate to look at number of shards to make conclusions around the local cluster being completed. See our docs (https://www.elastic.co/guide/en/elasticsearch/reference/current/search-search.html#search-api-response-body ): 

> Note that shards that are not allocated will be considered neither successful nor failed. Having failed+successful less than total is thus an indication that some of the shards were not allocated.

Could we simplify things here by relying on the specific callback that gets called once the local cluster is done with its part? I guess the challenge is that `onFinalReduce` and `onPartialReduce` both call `updatePartialResponse`. I think that we could mark the local cluster done when onFinalReduce is called and we are minimizing roundtrips.

javanna reviewed (commented):
left a couple of comments, great progress!!

quux00 commented on server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java:
```diff
@@ -402,6 +411,16 @@ void executeRequest(
+        if (localIndices.indices() == null) {
+            return false;
+        }
+        return localIndices.indices().length > 0;
```
OK. The original PR just used `localIndices == null` as the check for the absence of local indices. I expanded it to do all these additional checks just to be "safe", but sounds like that's the wrong move and I should go back to the single null check. I see now that it comes from:

```
remoteClusterIndices = remoteClusterService.groupIndices(rewritten.indicesOptions(), rewritten.indices());

OriginalIndices localIndices = remoteClusterIndices.remove(RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY);
```
so I believe the simple null check is sufficient based on that. I'll change it back.

quux00 committed:
PR feedback

javanna commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchTask.java:
```diff
@@ -403,6 +405,7 @@ protected void onFetchFailure(int shardIndex, SearchShardTarget shardTarget, Exc
         protected void onListShards(List<SearchShard> shards, List<SearchShard> skipped, Clusters clusters, boolean fetchPhase) {
             // best effort to cancel expired tasks
             checkCancellation();
+            ccsMinimizeRoundtrips = (clusters.isCcsMinimizeRoundtrips() && clusters.getRemoteClusters() > 0);
```
nit: do we need to set ccsMinimizeRoundtrips to true when there are no remote clusters? Maybe we could default  it to false instead?

quux00 replied:
I think that's what I am doing?

Technically the Clusters object should know whether we are doing minimizeRoundtrips, so we should just be able to do `ccsMinimizeRoundtrips = clusters.isCcsMinimizeRoundtrips()` but I decided to add the extra protection of ensure there is at least one remote cluster.

Actually, maybe the best solution is to put an assert in the new `Clusters` constructor that takes the new arguments:

```
assert (ccsMinimizeRoundtrips ? remoteClusters > 0 : true) : "ccsMinimizeRoundtrips is true but remoteClusters is " + remoteClusters;
```

Then we can just do `ccsMinimizeRoundtrips = clusters.isCcsMinimizeRoundtrips()` here. Let me know if that sounds best to you.

javanna replied:
++

javanna replied:
this sounds like a good idea to me.

quux00 replied:
Fixed in most recent commit.

javanna commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java:
```diff
@@ -84,13 +91,16 @@ class MutableSearchResponse {
      * Updates the response with the result of a partial reduction.
      * @param reducedAggs is a strategy for producing the reduced aggs
+     * @param isFinalLocalReduce true if the local cluster search has finished (during CCS, this can be true, but the overall
+     *                           search is still running on remote clusters)
```
add "in case roundtrips are minimized" ?

quux00 replied:
Good catch. Fixed.

javanna commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java:
```diff
@@ -103,18 +113,35 @@ synchronized void updatePartialResponse(
         this.reducedAggsSource = reducedAggs;
         this.reducePhase = reducePhase;
+        if (isFinalLocalReduce) {
+            if (localClusterSearchCompleted == false
```
is this flag still needed? I think you could have an assert that verifies that final reduction is performed once and remove the flag. Am I missing something?

quux00 replied:
Yeah, good question. I did try to remove it in my last PR rewrite, but I thought I had found it was still needed. But at your prompt, I removed it again just now and tested it again and now I don't find any case where it's needed, so I'll remove it in the next commit.

javanna commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java:
```diff
@@ -103,18 +113,35 @@ synchronized void updatePartialResponse(
+        if (isFinalLocalReduce) {
+            if (localClusterSearchCompleted == false
+                && clusters.getTotal() > 0
+                && clusters.getTotal() > (clusters.getSuccessful() + clusters.getSkipped())) {
```
can you help me understand this conditional? I don't remember why it's needed.

quux00 replied:
In all cases (that I know of) except `ccsMinimizeRoundtrips`, the Clusters object is created in its final state even at the start of an operation. By final state, I mean `successful + skipped = total`.

Examples: 

1) CCS NOT doing minimize_roundtrips it is created at the beginning with say `Clusters{total: 3; successful: 2; skipped: 1}`, so you don't want to increment successful here. 

2) In addition, there are cases where Clusters is initialized to Clusters.EMPTY (not totally sure why), but without this protection that causes those test cases to fail because you get `Clusters{total: 0; successful: 1; skipped: 0}` 

So this sanity check ensures that `successful` is incremented only in scenarios where the Cluster object has been set up to be incrementally updated (which right now only the ccsMinimizeRoundtrips scenario does).

quux00 replied:
I added additional commentary.

javanna reviewed (commented):
Looks like this is really close. I left a couple of nits and questions.

quux00 committed:
PR feedback 2

javanna commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java:
```diff
@@ -103,18 +112,32 @@ synchronized void updatePartialResponse(
         this.reducedAggsSource = reducedAggs;
         this.reducePhase = reducePhase;
+        if (isFinalLocalReduce && clusters.getTotal() > (clusters.getSuccessful() + clusters.getSkipped())) {
+            // currently only the ccsMinimizeRoundTrip=true creates Clusters in their initial state (where successful=0)
```
would it be more explicit to rely on the clusters.isMinimizeRoundtrips here instead of on the count of the clusters to guess whether the object needs updating or not?

quux00 replied:
For now that would probably work (would have to test to be sure of edge cases), but it would then lock in the functionality to only isMinimizeRoundtrips, but maybe that's better for now. I'll test it.

quux00 replied:
Looks good. Fixed.

quux00 committed:
PR feedback 3

javanna reviewed (approved):
I left one suggestion around the Clusters class changes that may help for safety, LGTM otherwise. Thanks for all the iterations!

quux00 commented on server/src/main/java/org/elasticsearch/action/search/SearchResponse.java:
```diff
@@ -479,6 +479,33 @@ public static class Clusters implements ToXContentFragment, Writeable {
+                : "total: " + total + " successful: " + successful + " skipped: " + skipped + " remote: " + remoteClusters;
+            assert successful <= total : "total: " + total + " successful: " + successful + " skipped: " + skipped;
+            assert remoteClusters <= total : "total: " + total + " remote: " + remoteClusters;
+            assert (ccsMinimizeRoundtrips ? remoteClusters > 0 : true)
```
Either is good. I changed it.

quux00 commented on server/src/main/java/org/elasticsearch/action/search/SearchResponse.java:
```diff
@@ -546,14 +591,22 @@ public boolean equals(Object o) {
             return total == clusters.total && successful == clusters.successful && skipped == clusters.skipped;
         }
 
+        // TODO: should this be updated to include the new fields?
```
I think we should wait until we have discussions with Tyler Perkins and the Kibana team about whether we are going to enhance the `_clusters` section in the REST response. It's possible these new fields will need to be written to the .async-search system index based on that decision. I think at that point we could look into extended the class for async search only.

Since this is an inner class of SearchResponse and Clusters is passed around a lot through the codebase (for async and sync search), that could be a fairly extensive refactoring to do, so my vote is to delay until we know what the long term plan is.

I'll add a note to the code about this.

javanna replied:
I see, I trust your judgement, I think I may have underestimated the effort needed. Let's mark the fields `transient` and ship it?!

quux00 replied:
Good idea on marking them `transient`

quux00 committed:
PR feedback 4

quux00 commented:
@elasticsearchmachine generate changelog

elasticsearchmachine commented:
Hi @quux00, I've created a changelog YAML for you.

quux00 committed:
Update docs/changelog/96012.yaml

javanna commented on server/src/main/java/org/elasticsearch/action/search/SearchResponse.java:
```diff
@@ -480,10 +479,27 @@ public static class Clusters implements ToXContentFragment, Writeable {
+        /**
+         * If finalState = true, creates a Clusters object in its "final" state, where successful + skipped = total
+         * If finalState = false, then the above condition is not enforced. This would happen in a case
+         * where an async search is happening and the Clusters state will be incrementally updated
```
I keep forgetting: this is only true for async search, or is it the same also for synchronous search when minimizing roundtrips?

quux00 replied:
Synchronous search will also use it because of the way that Async Search just layers on top of Sync Search - the logic for creating this new Cluster object is in TransportSearchAction which doesn't check whether it is an async search or not.

javanna commented on server/src/main/java/org/elasticsearch/action/search/SearchResponse.java:
```diff
@@ -557,6 +573,82 @@ public String toString() {
+     * Like Clusters, this is an immutable class.
+     *
+     * Unlike Clusters, (for now at least) the new fields are not Writeable. They are only
+     * used during CCS searches for accounting on the primary coordinator.
```
it may be worth mentioning async search here as in fact this info is only useful as part of async search in the coordinating node?

javanna commented on server/src/main/java/org/elasticsearch/action/search/SearchResponse.java:
```diff
@@ -557,6 +573,82 @@ public String toString() {
+     * Unlike Clusters, (for now at least) the new fields are not Writeable. They are only
+     * used during CCS searches for accounting on the primary coordinator.
+     */
+    public static class CcsClusters extends Clusters {
```
The point of having the Clusters class in the first place is to handle CCS, so to have one extension of Clusters that is called CCSClusters is kinda weird :)  Looking at its usage, this is a very specific extension that is only used when minimized roundtrips is on. How about `MinimizedRoundtripsAwareClusters` or something like that?

javanna commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchTask.java:
```diff
@@ -403,6 +405,7 @@ protected void onFetchFailure(int shardIndex, SearchShardTarget shardTarget, Exc
         protected void onListShards(List<SearchShard> shards, List<SearchShard> skipped, Clusters clusters, boolean fetchPhase) {
             // best effort to cancel expired tasks
             checkCancellation();
+            ccsMinimizeRoundtrips = clusters instanceof SearchResponse.CcsClusters ccsClusters && ccsClusters.isCcsMinimizeRoundtrips();
```
you can remove the instanceof checks by adding the isCcsMinimizeRoundtrips method to the base class and make it always return false?

javanna commented on server/src/main/java/org/elasticsearch/action/search/SearchResponse.java:
```diff
@@ -557,6 +573,82 @@ public String toString() {
+    /**
+     * Extension of the Clusters class for CCS uses. It adds the number of remote clusters
+     * included in a search and whether ccs_minimize_roundtrips is true.
+     *
```
not that it matters a lot, but IDEA gives me a warning that these blank lines will be ignored. Maybe we can remove them then?

javanna reviewed (approved):
I looked at the latest changes around the Clusters class and left some comments. I find that it does make it clearer when the additional fields are needed. I initially thought we could also use the subclass to enforce that it's never serialized but I don't think that is true. I am 50/50 on which approach I find better, both have pros and cons, either way is good with me.

quux00 committed:
Minor javadoc cleanup

quux00 commented on server/src/main/java/org/elasticsearch/action/search/SearchResponse.java:
```diff
@@ -480,10 +479,27 @@ public static class Clusters implements ToXContentFragment, Writeable {
+         * All clusters must be accounted for.
+         * @param total total number of clusters being searched
+         * @param successful
+         * @param skipped
```
Fixed.

javanna reviewed (approved):
SHIP IT

quux00 commented:
@abdonpijpelink - Would you have time to review the end user docs changes here?

You can look at this commit to see the new changes: https://github.com/elastic/elasticsearch/pull/96012/commits/a6460bd4087a9f1edb5935f1ae5ea8b9b26de140

quux00 commented:
@abdonpijpelink

I just talked with Luca and we are going to change the default setting in a follow-on PR, so the end user docs changes will go into that new PR so you can wait to review it then. I'll be deleting those commits from this PR in a moment. Sorry for the change.

abdonpijpelink reviewed (commented):
Sure thing! For your follow-up PR: note that the `<1>` numbers ("callouts") have to restart from 1 for each code snippet, otherwise the docs won't build.
